module: Add sched module

valschneider opened the pull request:
This module is a collection of scheduler data getters.

There's quite a few places in LISA where we re-implement these (mostly capacity getters), so I figured this would be the best place to bunch it up.

I'm a bit conflicted about the handling of the EAS energy model - we have a ~1000 line file that handles everything energy model related, and some of it could be moved here. However I don't think it's a simple task, so for now the handling of energy model data done in this module is trivial.

Also, I should make sure this runs fine on Python3 to get it in line with the other modules, but I figured I could collect some comments first :)

douglas-raillard-arm commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+            value = node
+            try:
+                value = int(node)
+            except:
```
Catch-all except are evil and must not be used apart in something like nosetest. That will even catch stuff like SyntaxError, which can be extremely confusing:

```
except ValueError:
```

douglas-raillard-arm commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+        }
+
+        for dest in set(packables.itervalues()):
+            setattr(self, dest, {})
```
It would be a better idea to store that in a dictionnary, and implement __getattr__ to use that dict:
* That allows inspection of what was read. When setting an attribute, it is not possible to distinguish these dynamic data attributes from methods and other things like that.
* That allows modifications after the object is initialized if necessary.
* That avoids a sysfs entry name shadowing a method of the object. With __getattr__, it is impossible to break existing code just by changing what sysfs hierarchy is read since real attributes have priority.

valschneider commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+                capacities[cpu] = self._get_dmips_capacity(cpu)
+            else:
+                self.logger.warning('Cannot read capacity for cpu{}, assuming 1024'.format(cpu))
+                capacities[cpu] = 1024
```
Yeah, I'm unsure about this one. One the one hand it handles the case of returning 1024 for targets that don't expose any of those nodes (which is what we want), on the other it will not scream if those nodes are not present on a target where we *want* them to be here.

valschneider commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+        """
+        :returns: The scheduler tick frequency on the target
+        """
+        return int(test_env.target.config.get('CONFIG_HZ'))
```
Woopsie daisy !

douglas-raillard-arm commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+        """
+        :returns: The scheduler tick frequency on the target
+        """
+        return int(test_env.target.config.get('CONFIG_HZ'))
```
That is actually calling .get() on a dict under the hood, so that still applies :)

valschneider replied:
Right, I'll clean that up to raise an exception if the `get()` gives me a `None`

douglas-raillard-arm replied:
As discussed offline, it may be even better to change config.get(key) signature to config.get(key,raise=False). When raise=True, it would throw an IndexError with a message saying that the config option does not exist on that kernel or some similar helpful message.

setrofim replied:
`strict` is preferred name for this flag, but yes, it could be added.

valschneider replied:
I'll add this - do we actually want it to default to True ? It screams API break but at the same time I think it's a saner default.

valschneider commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+
+    @staticmethod
+    def probe(target):
+        return True
```
So the thing is it actually depends on a few different things... For capacities on Android, you'd usually want to have an energy model. On Mainline, it's just a device tree entry and is separate from having EAS. And then the config is not always exposed, but that's yet another different dependency...

The sched domains in general will always be exposed, so you'll always be able to e.g. check the sched domain flags.

setrofim replied:
So *some* sort of test should definitely be here (you can query `target.os` to find out if target is running Android or Linux. It's up to you how through you wan to be with the test -- whether you want to check for everything or just a single key location which, if present, likely means the other stuff will be present too. There is a trade-off  here between the speed of the probe and risking missing something so that you end up failing much later than necessary.

douglas-raillard-arm commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+    re_node = re.compile(r"(?P<name>.*)(?P<digits>\d+)$")
+
+    @staticmethod
+    def _ends_with_digits(node):
```
I can't really find an established convention about leading underscore for names in classes, but at module level it means "don't export this in from XXX import *". If that is not part of the "user API", that qualifies for leading underscore even if used by different classes IMHO.

setrofim replied:
The convention is established by PEP8: https://www.python.org/dev/peps/pep-0008/#id47

> Use one leading underscore only for non-public methods and instance variables.

This is to do with hinting the accessibility of class members and nothing to do with imports. `from XXX import *` behavior is controlled by `__all__`: https://docs.python.org/2/tutorial/modules.html#importing-from-a-package

douglas-raillard-arm replied:
I also checked PEP8, non-public could be understood both ways, but it may be clarified in another part of the document I skipped.

PEP8 also adds:
> _single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import > objects whose name starts with an underscore.

Python 2.7.15 and 3.6.5 both behave like that by default, unless __all__ is defined.

setrofim replied:
Fair point, however that refers to top-level definitions (which *can* be imported). For a class member, it unambiguously means "internal to the class".

valschneider commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+
+        self._build_domains()
+
+class ProcFSNode(object):
```
I don't know if the whole packing mess makes sense outside of the `sched_domain` tree, but I don't think that would cause any major issue.

setrofim replied:
If thou think this is specific to `sched_domain`, then probably best to leave it here (though, in that case, i would change the name of the class to indicate the association).

setrofim commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+        """
+        cpus = range(0, self.target.number_of_cpus)
+
+        capacities = {cpu : 0 for cpu in cpus}
```
`{}` is preferred to `dict()`, but yes.

setrofim commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+    > 'MC'
+    """
+
+    sched_domain_root = '/proc/sys/kernel/sched_domain'
```
Please do not upper case.  Upper case should be used for constants defined at top level; here, it's just a class attribute, and should lower case.

setrofim commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,239 @@
+
+    @staticmethod
+    def _ends_with_digits(node):
+        if not isinstance(node, str):
```
From now on, we're trying to be Python 3 compatible, so please include `from past.builtins import basestring` at the top as well.

valschneider commented:
I *think* I addressed every comment, the major ones being:
* Added a test in `probe` to see if the sched domain procfs entries are there
* Changed the `setattr` bizz to a `dyn_attrs` dict + `__getattr__()`
* Added a `strict` flag for KernelConfig.get() (False by default)

valschneider commented:
Oh and before I forget - I still need to properly test this out with Python3

valschneider commented:
I got it to run with either Python2 or Python3. I'll start tweaking a scheduler test in LISA using this module to validate the interface is sane or if it's missing something before releasing it into the wild.

derkling commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,262 @@
+
+    name = 'sched'
+
+    sysfs_root = '/sys/devices/system/'
```
Maybe a bit confusing name? That's not the sysfs root... but the system's sysfs root...

derkling commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,262 @@
+    sysfs_root = '/sys/devices/system/'
+
+    @staticmethod
+    def probe(target):
```
Maybe too generic name? You are probing just for SDs, isn't it?

derkling replied:
And... since it's not used by this module, I guess it's an API moved here from some other LISA module... hence that's a good chance to have a better descriptive name.

derkling replied:
For example, such a method should not better be part of the SchedDomainData class?
Something like:
`SchedDomainData.exists()` or `SchedDomainData.valid()` or `SchedDomainData.available()` ?

valschneider replied:
Actually it's imposed by the devlib module API - it's run to see if it makes sense to load the module.

derkling commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,262 @@
+        return int(self.target.config.get('CONFIG_HZ', strict=True))
+
+
+class SchedDomainData(object):
```
This class is used by the above code... why not moving it at the beginning of the module?

derkling commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,262 @@
+
+        return len(target.list_directory(SchedDomainData.sched_domain_root)) > 0
+
+    def get_sd_info(self, cpu):
```
Don't see the need for such a "dummy" wrapper... is it just to avoid passing in target?
I would say clients can directly call `SchedDomainData(self.target, cpu)`, why not?

derkling commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,262 @@
+from past.builtins import basestring
+
+
+class SchedModule(Module):
```
Looking at the methods below this seems more like a CPU module... we are not really reporting properties of the scheduler but instead properties of CPUs... isn't it?

valschneider replied:
In a way yes, although it's the scheduler's view of the CPUs

derkling commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,262 @@
+        return capacities
+
+    @memoized
+    def get_hz(self):
```
Don't particularly like the usage of "get_" prefixes... is that to be consistent with other modules?
Otherwise, I would probably like better these to be defined like "properties"...

valschneider replied:
Yeah modules like thermal or cpufreq have loads of set/get methods, so I tried to follow that.

setrofim replied:
It should be either `get_hz` as it is now, or `hz` with an additional `@property` decorator (method names should always be verbs, but properties are data types).

derkling commented on devlib/module/sched.py:
```diff
@@ -0,0 +1,262 @@
+
+class SchedDomain(SchedProcFSNode):
+
+    # Domain flags obtained from include/linux/sched/topology.h on v4.17
```
Should we add some sort of sanity check mechanism to verify that the target's kernel matches a min,max version?
Could eventually be done in the future... but maybe splitting out a warning in case the kernel is not a v4.17 just to say that we are using the flags as per their definition in kernel v4.17?

valschneider commented:
So, turns out if you have a test that wants to go through all of the cpus' sched domain data, it takes quite some time (one `read_tree_value` per CPU). So I changed things to make it possible to do it all in one `read_tree_value` call, which led me to restructure the whole  `SchedProcFSNode` thing.

A simple changeset would be:
* Changed the current `get_sd_info` into `get_cpu_sd_info`
* Added a new `get_sd_info` method (for all CPUs)
* Added a target kernel version check at probe time to warn if it's too old compared to the version used to create the module

setrofim commented:
Is this now good to be merged, or are you still testing, @valschneider ?

valschneider commented:
@setrofim If you're happy with it I think it's good to go. The "only" contentious point so far is this module has some knowledge of the energy model, which could live in a module/class of its own. Thing is, it's a bit more work to get that in a good shape, so I think for now I'll stick with the current implementation and just make it query the EM class/module when I'll have more bandwidth.
